flipperdevices/flipperzero-firmware
# [FL-3669] Expansion module protocol

# What's new

- UART refactoring by @skotopes (better access model) 
- Standard expansion module protocol with automatic detection, software flow control and RPC support
- Reference parser implementation
- Integration test application and unit tests for the parser

# TODO:

- [x] Documentation (Doxygen, protocol description)
- [x] Robustness (make it resilient to garbage data and unexpected disconnects)
- [x] Basic error detection via checksum
- [x] Code cleanup and optimisation

# Verification 

1. Connect the hardware programmer with jumper wires so it does not interfere with pins 13 to 15.
2. Connect pins **13** to **16** and **14** to **15** with jumper wires.
3. Flash the firmware.
4. Launch the test application: `./fbt launch APPSRC=expansion_test`
5. An audovisual alert should play shortly.
6. Check that the test applicaton has exited and the respective RPC session was closed.
7. Download file `/ext/ExpansionTest/test.txt` using qFlipper or any other method. 
8. Check that it is identical to the file found in the application source `assets` folder.

# Checklist (For Reviewer)
